### PR TITLE
Graceful Shutdown for all services

### PR DESCRIPTION
**Pull request by xgreenx**
- Closes https://github.com/FuelLabs/fuel-core/issues/949
- Closes https://github.com/FuelLabs/fuel-core/issues/897

Removed second spawned task for `fuel_core_services::service::run` function.

Instead, wrapped it into `catch_unwind`.
Added a `shutdown` method for the task that allows doing some cleanup for methods. For, almost, all services, it does nothing, but can be reused in the future. Added handling of the shutdown for `Relayer` to act immediately instead of waiting for the synchronization's end. Improved handling of the stop signal for `fuel-core-sync`.

### CODE REVIEW

File: crates/services/sync/src/service/tests.rs [hunk 1]
```diff
@@ -60,12 +61,15 @@ async fn test_new_service() {
...
     );
+    let mut last_value = 0;
     while let Some(h) = rx.recv().await {
+        last_value = h;
         if h == 16 {
             break
         }
     }
 
+    assert_eq!(last_value, 16);
```
Comment (freesig) [inline crates/services/sync/src/service/tests.rs:72]: The loop only breaks on 16 so this is always true

File: crates/services/src/service.rs [hunk 2]
```diff
@@ -228,12 +236,13 @@ where
     tokio::task::spawn(
         async move {
             tracing::debug!("running");
-            let join_handler = run(service, stop_sender.clone());
+            let run = std::panic::AssertUnwindSafe(run(service, stop_sender.clone()));
             tracing::debug!("awaiting run");
-            let result = join_handler.await;
+            let result = run.catch_unwind().await;
```
Comment (freesig) [inline crates/services/src/service.rs:241]: Doesn't this panic need to be resumed?

**Review by freesig — CHANGES_REQUESTED** [review]

File: crates/services/sync/src/service/tests.rs (hunk 1 above)
Comment (xgreenx) [reply crates/services/sync/src/service/tests.rs:72]: Previously the loop didn't work=) It was `None` because the import task failed at the start.

Comment (freesig) [reply crates/services/sync/src/service/tests.rs:72]: ah I see

File: crates/services/src/service.rs (hunk 2 above)
Comment (xgreenx) [reply crates/services/src/service.rs:241]: It is useless because we have an anonymous task. But we can do that. Why not=)

**Review by Voxelot — APPROVED** [review]

File: crates/services/src/service.rs (hunk 2 above)
Comment (freesig) [reply crates/services/src/service.rs:241]: It just means drops are called for the task

**Review by freesig — APPROVED** [review]

**Pull request merged**